File: app/src/main/java/cn/ifafu/ifafu/data/http/parser/ScoreParser.java
package cn.ifafu.ifafu.data.http.parser;

import com.alibaba.fastjson.JSONObject;

import org.jetbrains.annotations.NotNull;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import cn.ifafu.ifafu.app.School;
import cn.ifafu.ifafu.entity.Score;
import cn.ifafu.ifafu.entity.User;
import cn.ifafu.ifafu.util.NumUtils;

public class ScoreParser extends BaseParser<List<Score>> {

    private String account;
    private String schoolCode;

    public ScoreParser(User user) {
        this.account = user.getAccount();
        this.schoolCode = user.getSchoolCode();
    }

    @Override
    public List<Score> parse(@NotNull String html) {
        List<Score> list = new ArrayList<>();
        Document document = Jsoup.parse(html);
        Elements elementsTemp = document.select("table[id=\"Datagrid1\"]");
        if (elementsTemp.size() == 0) {
            return Collections.emptyList();
        }
        Elements elements = elementsTemp.get(0).getElementsByTag("tr");
        switch (schoolCode) {
            case School.FAFU:
                for (int i = 1; i < elements.size(); i++) {
                    list.add(paresToScoreFAFU(elements.get(i).children().eachText()));
                }
                break;
            case School.FAFU_JS:
                for (int i = 1; i < elements.size(); i++) {
                    list.add(paresToScoreFAFUJS(elements.get(i).children().eachText()));
                }
                break;
        }
        Collections.sort(list, (o1, o2) -> Long.compare(o1.getId(), o2.getId()));
        for (Score score : list) {
            score.setAccount(account);
            score.setId(score.getId() * 31 + account.hashCode());
        }
        return list;
    }

    private Score paresToScoreFAFU(List<String> eles) {
        Score score = new Score();
        score.setYear(eles.get(0));
        score.setTerm(eles.get(1));
        score.setId(Long.parseLong(eles.get(2)));
        score.setName(eles.get(3));
        score.setNature(eles.get(4));
        score.setAttr(eles.get(5));
        if (!eles.get(6).isEmpty()) {
            score.setCredit(NumUtils.toFloat(eles.get(6)));
        }
        String ele7 = eles.get(7);
        if (ele7.contains("免修")) {
            score.setScore(Score.FREE_COURSE);
        } else {
            if (!eles.get(7).isEmpty()) {
                score.setScore(NumUtils.toFloat(eles.get(7)));
            }
        }
        System.out.println(JSONObject.toJSONString(eles));
        if (!eles.get(8).isEmpty()) {
            score.setMakeupScore(NumUtils.toFloat(eles.get(8)));
        }
        score.setRestudy(eles.get(9).isEmpty());
        score.setInstitute(eles.get(10));
        if (eles.size() > 13) {
            if (!eles.get(11).isEmpty()) {
                score.setGpa(NumUtils.toFloat(eles.get(11)));
            }
            score.setRemarks(eles.get(12));
            score.setMakeupRemarks(eles.get(13));
        } else {
            score.setRemarks(eles.get(11));
            score.setMakeupRemarks(eles.get(12));
        }
        score.setIESItem(score.getScore() != Score.FREE_COURSE
                && !score.getNature().contains("任意选修")
                && !score.getNature().contains("公共选修")
                && !score.getName().contains("体育"));
        return score;
    }

    private Score paresToScoreFAFUJS(List<String> eles) {
        Score score = new Score();
        score.setId(Long.parseLong(eles.get(2)));
        score.setYear(eles.get(0));
        score.setTerm(eles.get(1));
        score.setName(eles.get(3));
        score.setNature(eles.get(4));
        score.setAttr(eles.get(5));
        score.setCredit(NumUtils.toFloat(eles.get(6)));
        score.setGpa(NumUtils.toFloat(eles.get(7)));
        score.setScore(NumUtils.toFloat(eles.get(8)));
        score.setMakeupScore(NumUtils.toFloat(eles.get(10)));
        score.setInstitute(eles.get(12));
        try {
            score.setRemarks(eles.get(13));
            score.setMakeupRemarks(eles.get(14));
        } catch (Exception e) {
            e.printStackTrace();
        }
        score.setIESItem(!score.getNature().contains("任意选修") && !score.getName().contains("体育"));
        return score;
    }

}

File: app/src/main/java/cn/ifafu/ifafu/data/http/parser/ScoreParser.kt
package cn.ifafu.ifafu.data.http.parser

import cn.ifafu.ifafu.app.School
import cn.ifafu.ifafu.entity.Score
import cn.ifafu.ifafu.entity.User
import org.jsoup.Jsoup
import java.util.*

class ScoreParser(user: User) : BaseParser<List<Score>>() {
    private val account: String = user.account
    private val schoolCode: String = user.schoolCode

    override fun parse(html: String): List<Score> {
        check(html)
        val list: MutableList<Score> = ArrayList()
        val document = Jsoup.parse(html)
        val elementsTemp = document.select("table[id=\"Datagrid1\"]")
        if (elementsTemp.size == 0) {
            return emptyList()
        }
        val elements = elementsTemp[0].getElementsByTag("tr")
        val parse = when (schoolCode) {
            School.FAFU ->
                this::paresToScoreFAFU
            School.FAFU_JS ->
                this::paresToScoreFAFUJS
            else ->
                throw UnknownError("未知学校ID")
        }
        elements.drop(1).forEach {
            //Element.eachText会去除空字符串
            val eachText = it.children().text().split(" ")
            list.add(parse(eachText))
        }
        for (score in list) {
            score.account = account
            score.id = score.id * 31 + account.hashCode()
        }
        return list
    }

    private fun paresToScoreFAFU(eles: List<String>): Score {
        val score = Score()
        score.year = eles[0]
        score.term = eles[1]
        score.id = eles[2].toLong()
        score.name = eles[3]
        score.nature = eles[4]
        score.attr = eles[5]
        score.credit = eles[6].toFloatOrNull() ?: -1F
        val ele7 = eles[7]
        if (ele7.contains("免修")) {
            score.score = Score.FREE_COURSE
        } else {
            score.score = eles[7].toFloatOrNull() ?: -1F
        }
        score.makeupScore = eles[8].toFloatOrNull() ?: -1F
        score.restudy = eles[9].isEmpty()
        score.institute = eles[10]
        if (eles.size > 13) {
            score.gpa = eles[11].toFloatOrNull() ?: -1F
            score.remarks = eles[12]
            score.makeupRemarks = eles[13]
        } else {
            score.remarks = eles[11]
            score.makeupRemarks = eles[12]
        }
        return score
    }

    private fun paresToScoreFAFUJS(eles: List<String>): Score {
        val score = Score()
        score.id = eles[2].toLong()
        score.year = eles[0]
        score.term = eles[1]
        score.name = eles[3]
        score.nature = eles[4]
        score.attr = eles[5]
        score.credit = eles[6].toFloatOrNull() ?: -1F
        score.gpa = eles[7].toFloatOrNull() ?: -1F
        score.score = eles[8].toFloatOrNull() ?: -1F
        score.makeupScore = eles[10].toFloatOrNull() ?: -1F
        score.institute = eles[12]
        score.remarks = eles[13]
        score.makeupRemarks = eles[14]
        return score
    }

}